astropy/io/misc/hdf5.py

```diff
@@ -80,11 +84,12 @@ def read_table_hdf5(input, path=None):
     except ImportError:
         raise Exception("h5py is required to read and write HDF5 files")
 
+    input_save = input
```

Comment:
It would be good to provide a code comment explaining what is happening since the logic of this function (with recursive calls and code that can change `input` in place) is a little intricate.  My understanding is that `input_save` finally gets used below only when `input` is an HDF5 group.

One thing I noticed is that the API allows passing in an HDF5 `Dataset` directly, but if that happens then `input_save` will be a `Dataset` and this might cause an error?